lib/ddtrace/contrib/rest_client/request_patch.rb

```diff
@@ -0,0 +1,100 @@
+require 'ddtrace/ext/net'
+require 'ddtrace/ext/distributed'
+require 'ddtrace/propagation/http_propagator'
+
+module Datadog
+  module Contrib
+    module RestClient
+      # RestClient RequestPatch
+      module RequestPatch
+        REQUEST_TRACE_NAME = 'rest_client.request'.freeze
+
+        def self.included(base)
+          if Gem::Version.new(RUBY_VERSION) < Gem::Version.new('2.0.0')
+            base.class_eval do
+              alias_method :execute_without_datadog, :execute
+              remove_method :execute
+              include InstanceMethods
+            end
+          else
+            base.send(:prepend, InstanceMethods)
+          end
+        end
+
+        # Compatibility shim for Rubies not supporting `.prepend`
+        module InstanceMethodsCompatibility
+          def execute(&block)
+            execute_without_datadog(&block)
+          end
+        end
+
+        # InstanceMethods - implementing instrumentation
+        module InstanceMethods
+          include InstanceMethodsCompatibility unless Gem::Version.new(RUBY_VERSION) >= Gem::Version.new('2.0.0')
+
+          def execute(&block)
+            datadog_trace_request do |span|
+              if datadog_configuration[:distributed_tracing] && datadog_pin.tracer.enabled
+                Datadog::HTTPPropagator.inject!(span.context, processed_headers)
+              end
+
+              super(&block)
+            end
+          end
+
+          def datadog_tag_request(span)
+            uri = URI.parse(url)
+            span.resource = method.to_s.upcase
+            span.span_type = Ext::HTTP::TYPE
+            span.set_tag(Ext::HTTP::URL, uri.path)
+            span.set_tag(Ext::HTTP::METHOD, method.to_s.upcase)
+            span.set_tag(Ext::NET::TARGET_HOST, uri.host)
+            span.set_tag(Ext::NET::TARGET_PORT, uri.port)
+          end
+
+          def datadog_trace_request
+            span = datadog_pin.tracer.trace(REQUEST_TRACE_NAME, service: datadog_pin.service_name)
+
+            datadog_tag_request(span)
+
+            response = yield span
+
+            span.set_tag(Ext::HTTP::STATUS_CODE, response.code)
+            response
+          rescue ::RestClient::ExceptionWithResponse => e
+            span.set_error(e) if Ext::HTTP::ERROR_RANGE.cover?(e.http_code)
+            span.set_tag(Ext::HTTP::STATUS_CODE, e.http_code)
+
+            raise e
+            # rubocop:disable Lint/RescueException
+          rescue Exception => e
+            # rubocop:enable Lint/RescueException
+            span.set_error(e)
+
+            raise e
+          ensure
+            span.finish
+          end
+
+          def datadog_pin
+            @datadog_pin ||= begin
+              service = datadog_configuration[:service_name]
+              tracer = datadog_configuration[:tracer]
+
+              Datadog::Pin.new(
```

Comment:
Thinking harder on this, initializing a `Datadog::Pin` for every request might be a bad idea, since it might end up allocating memory that gets tossed out anyways, but at scale could be problematic. That considered, we do need to allow a `Pin` to be added and configured, such that a user could override the behavior for a specific request, if they so chose.

I think maybe we need to change the strategy a little here: above where we use these configuration values from the`Pin`, I think we should decompose and encapsulate each of those options, to use the `Pin` if it exists, or otherwise to fallback to the global configuration (only if the `Pin` does not exist, not if the `Pin` is simply missing a value.)

This might look something like:

```ruby
def datadog_trace_request
  datadog_tracer.trace(REQUEST_TRACE_NAME) do
    # Do some work...
  end
end

def datadog_tracer
  respond_to?(:datadog_pin) ? datadog_pin.tracer : datadog_configuration[:tracer]
end

def datadog_service_name
  respond_to?(:datadog_pin) ? datadog_pin.service_name : datadog_configuration[:service_name]
end

# Do not define #datadog_pin.
# If user does `Pin.new(service_name: 'custom-service').onto(request)`
# then the pin will become available. Otherwise we don't want to create a Pin.
# `Datadog.configure(request, service_name: 'custom-service')` will probably not work,
# because it will never add a pin to an object: only configure an existing one.
# See https://github.com/DataDog/dd-trace-rb/blob/master/lib/ddtrace/pin.rb
#     https://github.com/DataDog/dd-trace-rb/blob/master/lib/ddtrace/configuration/pin_setup.rb
#     https://github.com/DataDog/dd-trace-rb/blob/master/lib/ddtrace.rb
# def datadog_pin
#   nil
# end
```

Reply:
Lets work on this optimization in separate PR's as there are a few approaches around `Pin` Like `PinSetup` and `Datadog.configure` and it would be good to unify them in the future.

Reply:
You're right, we probably shouldn't modify the `Pin` or `PinSetup` classes in this PR. However, I don't think creating pins here, even temporarily, is a satisfactory solution, given the possible performance impact.

So I would suggest we drop pin support for now altogether, and use global configuration only instead, until we can get the groundwork in for a satisfactory multiplexing solution in place: either by having optimized the use of Pins, or by leveraging the new integration configuration to drive multiplexing instead.